test-data/unit/check-selftype.test

```diff
@@ -248,3 +248,46 @@ class A:
 class B(A):
     def __init__(self, arg: T) -> None:
         super(B, self).__init__()
+
+[case testSelfTypeNonsensical]
+# flags: --hide-error-context
+from typing import TypeVar, Type
+
+T = TypeVar('T', bound=str)
+class A:
+    def foo(self: T) -> T:   # E: The type of self 'builtins.str' is not a supertype its class '__main__.A'
+        return self
+
+    @classmethod
+    def cfoo(cls: Type[T]) -> T:  # E: The type of self 'Type[builtins.str]' is not a supertype its class 'Type[__main__.A]'
+        return cls()
+
+Q = TypeVar('Q', bound='B')
+class B:
+    def foo(self: Q) -> Q:
+        return self
+
+    @classmethod
+    def cfoo(cls: Type[Q]) -> Q:
+        return cls()
+    
+class C:
+    def foo(self: C) -> C: return self
+
+    @classmethod
+    def cfoo(cls: Type[C]) -> C:
+        return cls()
+
+class D:
+    def foo(self: str) -> str:  # E: The type of self 'builtins.str' is not a supertype its class '__main__.D'
```

Comment:
Outside a stub file, we could allow a function like this in a function body, assuming that the first argument is not named `self` (see discussion at #2374 and #16). However, the function shouldn't be usable as an instance method. I've heard about production code that relies on this. It's a questionable idiom, as a missing `self` is usually an error, and if we allow this idiom it can result in false negatives. 

@gvanrossum What do you think?